### PR TITLE
MaxHit plugin

### PR DESCRIPTION
**Pull request by bartvollebregt**
Discussions on discord (**Discord name: Cashman**)
OSRS Username: **C4shman**

Partially based on https://github.com/runelite/runelite/pull/2841/files

### Max Hit
- Drawing the damage in the equipment window
- Range damage
- Melee damage
  - Obsidian
  - Berserker necklace
  - Prayer bonuses
  - Dharok
- Magic damage
  - Normal Spellbook
  - Ancient magicks
  - Tridents
  - Magic equipment
- Slayer equipment
- Slayer Dart
- Salve amulet

### Code style
- Run code analysis
- Optimize imports
- Optimize class / method access
- Run code style checks
- Add copyright and license
- Replace item names with id's
- Write tests

### Code style
- Refactor Attackstyles after #6643 has been merged

## Won't implement for now
- Special attacks
- Maxhit against specific monsters (e.g. arclight vs demons, salve amulet vs undead) 
- Magic spells that are not being auto-casted
- Maxhit during Castle Wars
- Charge (for god spells), since the only way to track it is through chat messages

### CODE REVIEW

**Comment by Alexsuperfly** [comment]
obsidian armor and berserker necklace

**Comment by bartvollebregt** [comment]
@deathbeam @Alexsuperfly Could you do an initial review?
I've probably missed some edge cases but so far I think quite a lot is in there already.

File: runelite-client/src/main/java/net/runelite/client/plugins/maxhit/config/CustomFormulaConfig.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+		))))),
+		(client, calculator) ->
+		{
+			int currentHP = client.getBoostedSkillLevel(Skill.HITPOINTS);
+			int maxHP = client.getRealSkillLevel(Skill.HITPOINTS);
+			int lostHP = maxHP - currentHP;
+
+			double initialMaxHit = calculator.calculateDefault();
+
+			double multiplier = (1.0 + lostHP / 100.0 * maxHP / 100.0);
```
Comment (Alexsuperfly) [inline runelite-client/src/main/java/net/runelite/client/plugins/maxhit/config/CustomFormulaConfig.java]: why cant this just be `1.0 + lostHP/100`

Comment (bartvollebregt) [reply runelite-client/src/main/java/net/runelite/client/plugins/maxhit/config/CustomFormulaConfig.java]: The maximum HP seems to be important in the formula. For example: when your maxHP is 70 and you've lost 19, the multiplier will be 1.133. When your maxHP is 20 and you've lost 19 the multiplier will be 1.038. 

The only thing I'm uncertain about is whether there should be parenthesis or not. 
This is the current order: `(1.0 + ((lostHP / 100.0) * (maxHP / 100.0)))`

**Comment by fuzzy1450** [comment]
The plugin does not initialize correctly if the equipment menu is opened without boots, gloves, ring, or something in the ammo slot.
 
image
image
image

Equipping boots, gloves, ring, or ammo causes the plugin to function properly, even after removing these items. The plugin then functions properly, even after closing and reopening the equipment menu, but this process needs to be repeated after logout.


Equipping an item in any other slot while the plugin is not initialized properly results in this exception.

`
 14:51:04 [Thread-5] WARN  n.runelite.client.eventbus.EventBus - Uncaught exception in event subscriber
java.lang.ArrayIndexOutOfBoundsException: 9
	at net.runelite.client.plugins.maxhit.equipment.EquipmentHelper.wearsItem(EquipmentHelper.java:40)
	at net.runelite.client.plugins.maxhit.equipment.EquipmentHelper.lambda$wearsItem$1(EquipmentHelper.java:46)
	at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1359)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:449)
	at net.runelite.client.plugins.maxhit.equipment.EquipmentHelper.wearsItem(EquipmentHelper.java:45)
	at net.runelite.client.plugins.maxhit.equipment.EquipmentHelper.lambda$wearsItemSet$0(EquipmentHelper.java:35)
	at java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
	at java.util.Collections$2.tryAdvance(Collections.java:4717)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
	at java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.allMatch(ReferencePipeline.java:454)
	at net.runelite.client.plugins.maxhit.equipment.EquipmentHelper.wearsItemSet(EquipmentHelper.java:35)
	at net.runelite.client.plugins.maxhit.calculators.MaxHitCalculator.applyEquipmentBonus(MaxHitCalculator.java:108)
	at net.runelite.client.plugins.maxhit.calculators.MeleeMaxHitCalculator.calculate(MeleeMaxHitCalculator.java:128)
	at net.runelite.client.plugins.maxhit.calculators.MaxHitCalculator.getMaxHit(MaxHitCalculator.java:137)
	at net.runelite.client.plugins.maxhit.MaxHitPlugin.updateMaxHitWidget(MaxHitPlugin.java:85)
	at net.runelite.client.plugins.maxhit.MaxHitPlugin.onItemContainerChanged(MaxHitPlugin.java:56)
	at net.runelite.client.eventbus.EventBus$Subscriber.invoke(EventBus.java:72)
	at net.runelite.client.eventbus.EventBus.post(EventBus.java:218)
	at net.runelite.client.util.DeferredEventBus.replay(DeferredEventBus.java:69)
	at net.runelite.client.callback.Hooks.clientMainLoop(Hooks.java:146)
	at client.methodDraw(client.java:1362)
	at client.ak(client.java)
	at bf.ag(bf.java:397)
	at bf.run(bf.java:351)
	at java.lang.Thread.run(Thread.java:748)
`

**Comment by bartvollebregt** [comment]
@fuzzy1450 Thanks a lot for the testing!
I will make sure this is fixed as soon as possible

**Comment by bartvollebregt** [comment]
@fuzzy1450 It took me a while to reproduce it but got it eventually.
I've just pushed a fix which should resolve the problem, could you try again?

Btw: I've also updated my repo with the lastest runelite/master branch.

**Comment by Teifun2** [comment]
I sill have the Equip Your Character Message.
equip

**Comment by Nightfirecat** [comment]
This plugin would be [an excellent candidate for the plugin hub](/runelite/plugin-hub)

**Comment by Nightfirecat** [comment]
A max hit plugin [has been added to the plugin hub](https://runelite.net/plugin-hub/show/maxhitplugin). We encourage you to contribute to that one or, if for whatever reason that's not appropriate, create your own hub plugin.

**Pull request closed without merging**